### PR TITLE
[Feat][NSA] Implement nsa_forward_varlen kernel.

### PR DESCRIPTION
**Pull request by michaelwithu**
## Description

Solved #109 
This PR unifies the handling of fixed-length and variable-length sequences in the kernel by flattening the input from [B, T, H, D] to [B*T, H, D] and introducing an offsets array that marks sequence boundaries. This enables a single kernel implementation to support both cases.

For attention computation, we select ns chunks based on group query configuration and process them in a pipelined fashion. The pipeline interleaves memory loads of K/V blocks with computation phases—including QK GEMM, softmax, and value aggregation—to overlap data movement and arithmetic, improving throughput and memory efficiency.

## Fixed issues in the TileLang example

1. https://github.com/tile-ai/tilelang/blob/main/examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen.py#L96C18-L96C29

NS should be “ns = block_counts[bos + i_t, i_h]” which means the global index.

Moreover, when `s > 1` and `batch > 1`, it will also cause correctness issues.

2. https://github.com/tile-ai/tilelang/blob/main/examples/deepseek_nsa/example_tilelang_nsa_fwd_varlen.py#L72

batch means the real sequence number in varlen case
The term `c_seq_len` refers to the total number of tokens across all sequences.
Therefore, multiplying by the batch size is meaningless, because `c_seq_len` already represents the total number of tokens across all sequences.
Moreover, it would also lead to performance degradation.

## Type of Change

- [x] New operator implementation

## Checklist

- [x] I have run `pre-commit run --files “change files”` and fixed all linting issues.
- [x] I have verified that my changes pass local unit tests.
- [x] **(For new ops)** I have added the corresponding `Benchmark` class in `benchmarks/`.
- [x] **(For new ops)** I have reported benchmark results in the tracking issue.

### CODE REVIEW

File: top/kernels/deepseek_nsa/nsa_fwd.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                logsum = T.alloc_fragment([g], accum_dtype)
+
+                i_c, i_v, i_bh = bx, by, bz
+                _, i_h = i_bh // head_kv, i_bh % head_kv
+
+                i_n, i_t = token_indices[i_c, 0], token_indices[i_c, 1]
+
+                bos = offsets[i_n]
+
+                NS = block_counts[i_t, i_h]
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply top/kernels/deepseek_nsa/nsa_fwd.py]: You're such a little genius!

[the remark this answers is not in the document]
Comment (michaelwithu) [reply top/kernels/deepseek_nsa/nsa_fwd.py]: I've taken your advice.

File: benchmarks/deepseek_nsa/deepseek_nsa.py [hunk 2]
```diff
@@ -106,3 +118,286 @@ def baseline_profile(self,
...
+                block_indices[0, i, h, :len(i_i)] = i_i
+        block_indices = block_indices.sort(-1)[0]
+        block_counts = torch.randint(
+            1,
+            self.selected_blocks + 1,
+            (1, self.c_seq_len, self.head_kv),
+            dtype=torch.int32,
+            device="cuda",
+        )
+        token_indices = prepare_token_indices(offsets)
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply benchmarks/deepseek_nsa/deepseek_nsa.py]: I've taken your advice.

File: benchmarks/deepseek_nsa/deepseek_nsa.py [hunk 3]
```diff
@@ -106,3 +118,286 @@ def baseline_profile(self,
...
+        heads: int,
+        c_seq_len: int,
+        dim: int,
+        is_causal: bool,
+        scale: float,
+        block_size: int,
+        groups: int,
+        selected_blocks: int,
+        dtype: str,
+        accum_dtype: str,
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply benchmarks/deepseek_nsa/deepseek_nsa.py]: I've taken your advice.

File: top/kernels/deepseek_nsa/nsa_fwd.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+        self.selected_blocks = selected_blocks
+        self.dtype = dtype
+        self.accum_dtype = accum_dtype
+        self.dtype_name = str(dtype).split('.')[-1]
+        self.accum_dtype_name = str(accum_dtype).split('.')[-1]
+
+        self.kernel = _nsa_fwd_varlen_kernel(self.batch, self.heads, self.c_seq_len, self.dim,
+                                             self.is_causal, self.scale, self.block_size,
+                                             self.groups, self.selected_blocks, self.dtype,
+                                             self.accum_dtype)
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply top/kernels/deepseek_nsa/nsa_fwd.py]: I've taken your advice.

File: top/kernels/deepseek_nsa/nsa_fwd.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        block_counts_dtype = T.int32
+        offsets_dtype = T.int32
+        token_indices_dtype = T.int32
+
+        block_s = block_size
+        block_t = min(128, tilelang.math.next_power_of_2(dim))
+
+        nk = tilelang.cdiv(dim, block_t)
+        nv = tilelang.cdiv(dim, block_t)
+        assert nk == 1, "The key dimension can not be larger than 256"
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply top/kernels/deepseek_nsa/nsa_fwd.py]: solved.

File: benchmarks/deepseek_nsa/deepseek_nsa.py [hunk 6]
```diff
@@ -106,3 +118,286 @@ def baseline_profile(self,
...
+        block_indices_memory = self.head_kv * self.c_seq_len * self.selected_blocks
+        return (q_memory + k_memory + v_memory + output_memory + block_indices_memory)
+
+    def gen_inputs(self) -> tuple[torch.Tensor]:
+        offsets = (
+            torch.cat(
+                [
+                    torch.tensor([0], dtype=torch.long),
+                    torch.arange(
+                        16, self.c_seq_len)[torch.randperm(self.c_seq_len - 1)[:self.batch - 1]],
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply benchmarks/deepseek_nsa/deepseek_nsa.py]: solved.

File: benchmarks/deepseek_nsa/deepseek_nsa.py [hunk 7]
```diff
@@ -106,3 +118,286 @@ def baseline_profile(self,
...
+                s_i = s_b[i_q] if isinstance(block_counts, torch.Tensor) else s_b
+                k_i_slc, v_i_slc = (
+                    x.gather(0,
+                             i_i.clamp(0, t - 1).unsqueeze(-1).expand(*i_i.shape, x.shape[-1]))
+                    for x in (k_b, v_b))
+                # [s*bs, hq]
+                attn_slc = (
+                    torch.einsum("h d, n h d -> n h", q_i, k_i_slc).masked_fill(
+                        torch.logical_or(i_i < 0, i_i > i_q)
+                        | (c >= s_i if block_counts is not None else False),
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply benchmarks/deepseek_nsa/deepseek_nsa.py:326]: Focus the current code on causal attention only.

File: top/kernels/deepseek_nsa/nsa_fwd.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+                T.fill(acc_o, 0)
+                T.fill(logsum, 0)
+                T.fill(scores_max, -T.infinity(accum_dtype))
+
+                for i in T.Pipelined(ns, num_stages=num_stages):
+                    i_s = block_indices[bos + i_t, i_h, i] * bs
+                    if i_s <= i_t and i_s >= 0:
+                        # [BS, BK]
+                        # Lei: may have some padding issues
+                        # we should learn from mha varlen templates to handle this
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply top/kernels/deepseek_nsa/nsa_fwd.py]: solved.

File: tests/ops/test_deepseek_nsa_fwd.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    """Set up the test environment."""
+    torch.manual_seed(1234)
+
+
+@pytest.mark.parametrize(
+    ("batch, heads, c_seq_len, dim, is_causal, scale, block_size, "
+     "groups, selected_blocks, dtype, accum_dtype, tune"),
+    [
+        (1, 16, 1024, 64, True, 0.1, 32, 16, 1, torch.float16, torch.float32, False),
+        (4, 16, 8192, 64, True, 0.1, 32, 16, 1, torch.float16, torch.float32, False),
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply tests/ops/test_deepseek_nsa_fwd.py:21]: According to the current framework, testing can only be performed through a `main()` function; pytest is reserved for future use.

File: top/ops/deepseek_nsa.py [hunk 10]
```diff
@@ -66,3 +67,60 @@ def forward(
...
+            self.c_seq_len,
+            self.dim,
+            self.is_causal,
+            self.scale,
+            self.block_size,
+            self.groups,
+            self.selected_blocks,
+            self.dtype,
+            self.accum_dtype,
+            tune=self.tune)
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply top/ops/deepseek_nsa.py]: solved.

File: top/kernels/deepseek_nsa/nsa_fwd.py [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+                T.fill(logsum, 0)
+                T.fill(scores_max, -T.infinity(accum_dtype))
+
+                for i in T.Pipelined(ns, num_stages=num_stages):
+                    i_s = block_indices[bos + i_t, i_h, i] * bs
+                    if i_s <= i_t and i_s >= 0:
+                        # [BS, BK]
+                        # TODO(TileOPs): Handle boundary conditions if i_s + bs > current_seq_len
+                        # we should learn from mha varlen templates to handle this
+                        T.copy(k[bos + i_s:bos + i_s + bs, i_h, :bk], k_shared)
```
[the remark this answers is not in the document]
Comment (michaelwithu) [reply top/kernels/deepseek_nsa/nsa_fwd.py]: Retain the current implementation of the TileLang example kernel. Although testing has revealed performance issues, the TODO tasks will be addressed together with performance improvements in a later optimization phase.

File: benchmarks/deepseek_nsa/deepseek_nsa.py [hunk 12]
```diff
@@ -1,10 +1,22 @@
...
+
+if parse(fla.__version__) < parse("0.2.1"):
+    from fla.ops.common.utils import prepare_token_indices
+else:
+    from fla.ops.utils import prepare_token_indices
+
+from typing import Union
+
+from einops import rearrange, repeat
+from native_sparse_attention.ops.parallel import parallel_nsa_fwd
```
Comment (RMLYC) [inline benchmarks/deepseek_nsa/deepseek_nsa.py]: import `native_sparse_attention` in function

**Review by superAngGao — APPROVED** [review]
This PR adds test files and benchmark code for the NSA. Moreover, it provides the var_len-GQA section of the NSA kernel.

File: benchmarks/deepseek_nsa/deepseek_nsa.py [hunk 13]
```diff
@@ -106,3 +118,287 @@ def baseline_profile(self,
...
+        # q, k, v, output, block_indices, block_counts, offsets, token_indices
+        # ignore block counts, offsets and token_indices memory
+        q_memory = self.heads * self.c_seq_len * self.dim * self.dtype.itemsize
+        k_memory = self.head_kv * self.c_seq_len * self.dim * self.dtype.itemsize
+        v_memory = self.head_kv * self.c_seq_len * self.dim * self.dtype.itemsize
+        output_memory = self.heads * self.c_seq_len * self.dim * self.dtype.itemsize
+        block_indices_memory = self.head_kv * self.c_seq_len * self.selected_blocks
+        return (q_memory + k_memory + v_memory + output_memory + block_indices_memory)
+
+    def gen_inputs(self) -> tuple[torch.Tensor]:
```
Comment (RMLYC) [inline benchmarks/deepseek_nsa/deepseek_nsa.py]: Error in function return data type annotation

File: benchmarks/deepseek_nsa/deepseek_nsa.py (hunk 12 above)
Comment (michaelwithu) [reply benchmarks/deepseek_nsa/deepseek_nsa.py]: solved.

File: benchmarks/deepseek_nsa/deepseek_nsa.py (hunk 13 above)
Comment (michaelwithu) [reply benchmarks/deepseek_nsa/deepseek_nsa.py]: solved.
